src/CollectionsOfViews.jl

```diff
@@ -0,0 +1,153 @@
+        return new{T, N}(indices, data, lin_idx)
+    end
+end
+
+# AbstractArray interface (https://docs.julialang.org/en/v1/manual/interfaces/#man-interface-array)
+Base.size(cv::ArrayOfVectorViews) = size(cv.lin_idx)
+@inline function Base.getindex(cv::ArrayOfVectorViews, linear_index::Int)
+    @boundscheck checkbounds(cv.lin_idx, linear_index)
+    return @inbounds view(cv.data, cv.indices[linear_index]:(cv.indices[linear_index+1]-1))
+end
+@inline function Base.getindex(cv::ArrayOfVectorViews, idx...)
+    linear_index = getindex(cv.lin_idx, idx...)
+    return @inbounds getindex(cv, linear_index)
+end
+Base.IndexStyle(::Type{<:ArrayOfVectorViews{<:Any, N}}) where N = Base.IndexStyle(Array{Int, N})
+
+# Constructors
+"""
+    ArrayOfVectorViews(f!::Function, data::Vector{T}, dims::NTuple{N, Int}; sizehint)
+
+Create an `ArrayOfVectorViews` to store many vector views of potentially different sizes,
+emulating an `Array{Vector{T}, N}` with size `dims`. However, it avoids allocating each vector individually
+by storing all data in `data`, and instead of `Vector{T}`, the each element is a `typeof(view(data, 1:2))`.
+
+When the length of each vector is unknown, the `ArrayOfVectorViews` can be created reasonably efficient
+with the following do-block, which creates an intermediate `buffer::ConstructionBuffer` supporting the
+[`push_at_index!`](@ref) function.
+```
+vector_views = ArrayOfVectorViews(data, dims; sizehint) do buffer
+    for (ind, val) in some_data
+        push_at_index!(buffer, val, ind)
+    end
+end
+```
+`sizehint` tells how much space to allocate for the index `ind` if no `val` has been added to that index before,
+or how much more space to allocate in case all previously allocated space for `ind` has been used up.
+"""
+function ArrayOfVectorViews(f!::F, data::Vector, dims::Tuple; sizehint = nothing) where {F <: Function}
+    sizehint === nothing && error("Providing sizehint is mandatory")
+    b = ConstructionBuffer(data, dims, sizehint)
+    f!(b)
+    return ArrayOfVectorViews(b)
+end
+
+"""
+    ArrayOfVectorViews(b::CollectionsOfViews.ConstructionBuffer)
+
+Creates the `ArrayOfVectorViews` directly from the `ConstructionBuffer` that was manually created and filled.
+"""
+function ArrayOfVectorViews(b::ConstructionBuffer{T}) where T
+    indices = Vector{Int}(undef, length(b.indices) + 1)
+    lin_idx = LinearIndices(b.indices)
+    data_length = sum(ar.ncurrent for ar in b.indices; init=0)
+    data = Vector{T}(undef, data_length)
+    data_index = 1
+    for (idx, ar) in pairs(b.indices)
+        copyto!(data, data_index, b.data, ar.start, ar.ncurrent)
+        indices[lin_idx[idx]] = data_index
+        data_index += ar.ncurrent
+    end
+    indices[length(indices)] = data_index
+    resize!(b.data, 0); sizehint!(b.data, 0)
+    isa(b.indices, Vector) && (resize!(b.indices, 0); sizehint!(b.indices, 0))
```

Comment:
Can we add some comment here why everything is set to zero length?

Reply:
Does this comment explain it well enough? 
```suggestion
    # Since user-code in "push_at_index!" have access to `b`, setting dimensions to zero here
    # allows GC:ing the data in `b` even in cases when the compiler cannot guarantee that 
    # it is unreachable.
    resize!(b.data, 0); sizehint!(b.data, 0)
    isa(b.indices, Vector) && (resize!(b.indices, 0); sizehint!(b.indices, 0))
```